### PR TITLE
Variable trimming for filter-abund-single.py

### LINKED ISSUE #1382: filter-abund-single.py should support variable coverage trimming
See the -V and -Z parameters on filter-abund

### PR DESCRIPTION
**Pull request by betatim**
Fixes #1382

- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make clean diff-cover` If it introduces new functionality in
  `scripts/` is it tested?
- [x] `make format diff_pylint_report cppcheck doc pydocstyle` Is it well
  formatted?
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Is it documented in the `ChangeLog`?
  http://en.wikipedia.org/wiki/Changelog#Format
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?
- [x] Do the changes respect streaming IO? (Are they
  tested for streaming IO?)
- [x] Is the Copyright year up to date?

Using https://gist.github.com/betatim/93a68b55a4d56f62df8393909a727102 to make the following plot:

download 30

Need to think briefly why the green is not perfectly on top of green.

### CODE REVIEW

**Comment by standage** [comment]
Test this please jenkins

**Comment by standage** [comment]
By the way @ctb, I don't think I have admin privileges to the khmer repo, but we should give @betatim write access to the repo (and anything else that may be needed for his contributions to trigger CI builds automatically).

**Comment by ctb** [comment]
You now have admin privs.

**Comment by ctb** [comment]
And you should enable 2FA :)

**Comment by betatim** [comment]
Done.

**Comment by betatim** [comment]
> Need to think briefly why the green is not perfectly on top of green.

Default kmer size in `filter-abund-single.py` is 32 and not 21.

With
```shell
$ load-into-counting.py -M 1e6 -k 21 -q temp/mixed-reads.kh data/mixed-reads.fa.gz
$ filter-abund.py -q -C 3 -V -Z 20 temp/mixed-reads.kh data/mixed-reads.fa.gz -o temp/mixed-trimV.fa
$ abundance-dist-single.py -q -s -M 1e6 -k 21 temp/mixed-trimV.fa temp/mixed-trimV.hist

$ filter-abund-single.py -q -M 1e6 -k 21 -C 3 -V -Z 20 data/mixed-reads.fa.gz -o temp/mixed-trimV-single.fa
$ abundance-dist-single.py -q -s -M 1e6 -k 21 temp/mixed-trimV-single.fa temp/mixed-trimV-single.hist
```
you now get:
image

**Comment by betatim** [comment]
Test this please jenkins

**Comment by standage** [comment]
Test this please Jenkins

**Comment by betatim** [comment]
Ready for review! @luizirber, @camillescott, @standage , or @ctb

File: tests/test_filter_abund.py [hunk 1]
```diff
@@ -228,16 +228,18 @@ def test_filter_abund_4_retain_low_abund():
 
     counting_ht = _make_counting(infile, K=17)
 
-    script = 'filter-abund.py'
-    args = ['-V', counting_ht, infile]
-    utils.runscript(script, args, in_dir)
+    for script, args in (('filter-abund.py',
```
Comment (ctb) [inline tests/test_filter_abund.py]: could you split this into two tests, please? we try to keep the tests as free of for loops and if statements as possible (i.e. in declarative mode) for simplicity purposes.

**Comment by ctb** [comment]
LGTM, other than the test comment - thx!

**Comment by ctb** [comment]
LGTM. What's the current status of CI - can I just merge this or should I go check everything on Mac OS X?

**Comment by standage** [comment]
Everything should be good on OS X now. I double-checked the console logs for the last build, and they show success. Proceed at will!

**Comment by ctb** [comment]
There's a missing checkbox - streaming i/o. Intentional?

**Comment by ctb** [comment]
See test_streaming_io.py; probably should add a test in there.

**Comment by betatim** [comment]
Didn't tick it because I didn't add/detract from the streaming-ness (I think).

Tried to add a test and then started thinking: how do the other scripts that build a countgraph and use it all in one go deal with streaming IO? `filter-abund-single.py` makes two passes over the data right now: build the CG, then filter. Where as `filter-abund.py` uses a CG that you made earlier.

**Comment by ctb** [comment]
Yes, I was confused :). Streaming would require a holding cell for the data. Trim-low-abund is the script to use for streaming.

--
Titus Brown, ctbrown@ucdavis.edu

**Comment by betatim** [comment]
I'll tick it then and make it merge-able.

**Pull request merged**